python-checks/src/main/java/org/sonar/python/checks/IdentityComparisonWithNewObjectCheck.java

```diff
@@ -0,0 +1,155 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.python.checks;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.IsExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Token;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+import static java.util.Arrays.asList;
+
+// https://jira.sonarsource.com/browse/RSPEC-5796
+// https://jira.sonarsource.com/browse/SONARPY-674
+@Rule(key = "S5796")
+public class IdentityComparisonWithNewObjectCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE_IS = "Replace this \"is\" operator with \"==\".";
+  private static final String MESSAGE_IS_NOT = "Replace this \"is not\" operator with \"!=\".";
+  private static final String MESSAGE_SECONDARY = "This expression creates a new object every time.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.IS, IdentityComparisonWithNewObjectCheck::checkIsComparison);
+  }
+
+  private static void checkIsComparison(SubscriptionContext subscriptionContext) {
+    final IsExpression isExpr = (IsExpression) subscriptionContext.syntaxNode();
+    // The `if` merely ensures that an issue is reported at most once per operator.
+    if (!checkOperand(isExpr.leftOperand(), isExpr, subscriptionContext)) {
+      checkOperand(isExpr.rightOperand(), isExpr, subscriptionContext);
+    }
+  }
+
+  /**
+   * Checks a single operand of an <code>is/is not</code>-comparison.
+   * Returns <code>true</code> if it finds an issue.
+   */
+  private static boolean checkOperand(Expression operand, IsExpression isExpr, SubscriptionContext ctx) {
+    Optional<List<Tree>> secondariesOpt = findIssueForOperand(operand);
+    secondariesOpt.ifPresent(secondaryPositions -> {
```

Comment:
Nitpick but I'd rename `secondaryPositions` with `secondaryLocations` for consistency with our lexicon.